### ISSUE TITLE
(maybe a bug) weird training behavior

### ISSUE THREAD

**Issue by nicolaspanel** [mozilla/DeepSpeech#1953]
I'm using https://github.com/mozilla/DeepSpeech/commit/382707388ab32e9066c61dcaf621982942139500

Following picture show `training` and `validation` losses over multiple runs (ie multiple restarts reusing last checkpoint):
![Dessin sans titre (1)](https://user-images.githubusercontent.com/2500584/54302687-acb41b00-45c1-11e9-9170-bf51b66a63f9.png)

_losses are the one printed in the shell during training_

_`train` and `dev` sets are the same for runs 2 to 5_

**Observations**:
 - the first epoch of each run has a much slower `validation` loss
 - there is a huge difference between 2 successive epochs and this pattern repeat again and again

Do you have any idea of what could explain this ?

**Comment by nicolaspanel** [comment]
some additional details:
 - my dev set is pretty small (~2000 samples)
 - my dev does not match exactly the train set distribution. `dev` set records are shorter and are more noisy.

trainning logs:
```
****previous outputs removed****

Training of Epoch 18 - loss: 10.782960
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:02 Time:  0:33:02
I Validating epoch 18...
I Validation of Epoch 18 - loss: 44.131589
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 19...
I Training of Epoch 19 - loss: 10.719870
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:00 Time:  0:33:00
I Validating epoch 19...
I Validation of Epoch 19 - loss: 30.047718
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 20...
I Training of Epoch 20 - loss: 10.510714
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:00 Time:  0:33:00
I Validating epoch 20...
I Validation of Epoch 20 - loss: 44.507151
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 21...
I Training of Epoch 21 - loss: 10.487166
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:03 Time:  0:33:03
I Validating epoch 21...
I Validation of Epoch 21 - loss: 31.637611
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 22...
I Training of Epoch 22 - loss: 10.505370
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:00 Time:  0:33:00
I Validating epoch 22...
I Validation of Epoch 22 - loss: 45.537667
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 23...
I Training of Epoch 23 - loss: 10.350520
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:00 Time:  0:33:00
I Validating epoch 23...
I Validation of Epoch 23 - loss: 31.660463
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 24...
I Training of Epoch 24 - loss: 10.268184
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:01 Time:  0:33:01
I Validating epoch 24...
I Validation of Epoch 24 - loss: 47.216556
I Early stop triggered as (for last 10 steps) validation loss: 47.216556 with standard deviation: 8.930504 and mean: 35.083293
I FINISHED Optimization - training time: 5:30:46
100% (24 of 24) |###############################################| Elapsed Time: 0:00:02 Time:  0:00:02

****hide test details****
****start new run****

root@xxx:/DeepSpeech# python DeepSpeech.py  [...]
('Preprocessing', ['/data/train.csv'])
top('Loaded from cache at', '/data/train_cache.hdf5')
('Preprocessing', ['/data/cy_2019-03-12_dev_stt.csv'])
Preprocessing done
I STARTING Optimization
I Training epoch 25...
I Training of Epoch 25 - loss: 8.752240
100% (5115 of 5115) |###########################################| Elapsed Time: 0:32:55 Time:  0:32:55
I Validating epoch 25...
I Validation of Epoch 25 - loss: 19.319501
100% (24 of 24) |###############################################| Elapsed Time: 0:00:02 Time:  0:00:02
I Training epoch 26...
I Training of Epoch 26 - loss: 8.332165
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:03 Time:  0:33:03
I Validating epoch 26...
I Validation of Epoch 26 - loss: 47.577795
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 27...
I Training of Epoch 27 - loss: 8.070539
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:03 Time:  0:33:03
I Validating epoch 27...
I Validation of Epoch 27 - loss: 31.842970
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 28...
I Training of Epoch 28 - loss: 7.798905
100% (5115 of 5115) |###########################################| Elapsed Time: 0:33:03 Time:  0:33:03
I Validating epoch 28...
I Validation of Epoch 28 - loss: 48.417005
100% (24 of 24) |###############################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 29...
I Training of Epoch 29 - loss: 7.641581
100% (5115 of 5115) |######################################| Elapsed Time: 0:33:03 Time:  0:33:03
I Validating epoch 29...
I Validation of Epoch 29 - loss: 33.384423
100% (24 of 24) |##########################################| Elapsed Time: 0:00:03 Time:  0:00:03
I Training epoch 30...
I Training of Epoch 30 - loss: 7.497557
100% (5115 of 5115) |######################################| Elapsed Time: 0:33:03 Time:  0:33:03
I Validating epoch 30...
I Validation of Epoch 30 - loss: 48.733190
I Early stop triggered as (for last 6 steps) validation loss: 48.733190 with standard deviation: 10.867657 and mean: 36.108339
I FINISHED Optimization - training time: 3:18:32
100% (24 of 24) |##########################################| Elapsed Time: 0:00:02 Time:  0:00:02

****hide test details****
****start new run****

root@xxx:/DeepSpeech# python DeepSpeech.py   [...]
('Preprocessing', ['/data/train.csv'])
('Loaded from cache at', '/data/train_cache.hdf5')
('Preprocessing', ['/data/cy_2019-03-12_dev_stt.csv'])
Preprocessing done
I STARTING Optimization
I Training epoch 31...
I Training of Epoch 31 - loss: 6.244256
100% (5115 of 5115) |######################################| Elapsed Time: 0:32:54 Time:  0:32:54
I Validating epoch 31...
I Validation of Epoch 31 - loss: 20.581001
100% (24 of 24) |##########################################| Elapsed Time: 0:00:02 Time:  0:00:02
I Training epoch 32...
 17% (911 of 5115) |######                                 | Elapsed Time: 0:01:59 ETA:   0:10:39
 85% (4384 of 5115) |################################      | Elapsed Time: 0:23:02 ETA:   0:07:42 85% (4385 of 5115) |################################      | Elapsed Time: 0:23:05 ETA:   0:29:54I Training of Epoch 32 - loss: 5.884225
100% (5115 of 5115) |######################################| Elapsed Time: 0:34:19 Time:  0:34:19
I Validating epoch 32...
I Validation of Epoch 32 - loss: 52.134761
100% (24 of 24) |##########################################| Elapsed Time: 0:00:04 Time:  0:00:04
I Training epoch 33...
I Training of Epoch 33 - loss: 5.564034
100% (5115 of 5115) |######################################| Elapsed Time: 0:34:33 Time:  0:34:33
I Validating epoch 33...
I Validation of Epoch 33 - loss: 39.296744
100% (24 of 24) |##########################################| Elapsed Time: 0:00:03 Time:  0:00:03
```

**Comment by kdavis-mozilla** [comment]
I feel it's hard to conclude anything from the information presented.

For example there are many questions you leave un-answered...

1. Were the sm and md training sets drawn from the same distribution
2. Were the sm an md dev sets drawn from the same distribution
3. Why are you switching batch size between sm and md
4. Why within a run of sm or md does the validation error oscillate (On our data this never happens to this extent)
5. Why are you switching the learning rate between runs
6. Were the sm training and dev sets drawn from the same distribution
7. Were the md training and dev sets drawn from the same distribution
8...

Just saw that in the second post you answer the train vs dev distribution question for (both?) data sets when you say

> dev does not match exactly the train set distribution. dev set records are shorter and are more noisy.

Why are you doing this? This does not make sense to me.

**Comment by nicolaspanel** [comment]
@kdavis-mozilla thank you for your answer

> I feel it's hard to conclude anything from the information presented.
> 
> For example there are many questions you leave un-answered...
> 
> 1. Were the sm and md training sets drawn from the same distribution

see details below

> 2. Were the sm an md dev sets drawn from the same distribution

see details below

> 3. Why are you switching batch size between sm and md

The original goal was to speedup training since I switched from a 12G GPU to a 16G (I wasn't expected such improvement on both training and validation loss)


> 5. Why are you switching the learning rate between runs

I was just testing

> 6. Were the sm training and dev sets drawn from the same distribution

see below

> 7. Were the md training and dev sets drawn from the same distribution
>    8...

see below

> 
> > dev does not match exactly the train set distribution. dev set records are shorter and are more noisy.
> 
> Why are you doing this? This does not make sense to me.

My task focus on conversational audio (noisy / badly pronounced / etc.). This is why I use it exclusively for `dev`/`test` evaluations.  This come from the fact that I do not have enough annotated conversational data (~20k exemples, ~15h) . I then have to use data from other sources.


**dataset description**:
 - md dataset is an extension of sm dataset using data-augmentation (increased tempo for exemple). Goal is to reduce variance through more data
 - train: 330k samples. 20% [CV data](https://voice.mozilla.org) / 50%  [TS data](https://github.com/nicolaspanel/TrainingSpeech) (audio books, very low number of speaker) / 30% conversational speech
   durations distributions
   ![train_durations](https://user-images.githubusercontent.com/2500584/54345586-9a7ac100-4643-11e9-9aa8-bf0ae6a6154a.png)

 - dev: 2k samples. 100% conversational (sampled randomly from entire dataset)
   durations distribution
   ![dev_durations](https://user-images.githubusercontent.com/2500584/54343106-59cc7900-463e-11e9-8037-786dc8ec400b.png)

 - test: 2k samples. 100% conversational (sampled randomly from entire dataset)
   durations distribution
   ![test_durations](https://user-images.githubusercontent.com/2500584/54343166-82547300-463e-11e9-8db1-d70615ecd27c.png)


NOTABENE: 

> 4. Why within a run of sm or md does the validation error oscillate (On our data this never happens to this extent)



**My guess is:**
As far as I understand data feeding mechanism, batches are always run in the same order (short recordings first)

since TS has "longer" audio recordings than other source, the batches that are processed last (from 8 to 10 seconds audio) are ~85% from [TS data](https://github.com/nicolaspanel/TrainingSpeech)  (audiobooks), ~5% from [CV data](https://voice.mozilla.org) and ~10 from "conversational" dataset

Training is then very focus on working well on audiobooks data, which is not the goal here.

One option to prevent this could be to shuffle batches order while keeping recording of the same size together to prevent unecessary zero-padding.

Another option could be to remove 8 to 10 second data from TS to ensure the end of the training focus on conversational data…

**Comment by nicolaspanel** [comment]
BTW:
I have no idea what cause the `high validation loss` - `low validation loss` - `high validation loss` - `low validation loss` - ... pattern  or why first epoch of each run is always better than others

**Comment by kdavis-mozilla** [comment]
@nicolaspanel I think as you've such a complicated setup, different data sets, different learning rates, different GPU's, different batch sizes.... to actually determine if this is or is not a bug.

Is there anyway you can create a much simpler example that exhibits this effect? Maybe just on Common Voice data so we can rule out some unknowns.

**Comment by nicolaspanel** [comment]
@kdavis-mozilla I will look into it but shuffling batches could be a good idea anyway.

@ruben what do you think ?

**Comment by nicolaspanel** [comment]
@kdavis-mozilla  I'm not sure I can reproduce with another dataset since it seems to be related to difference of distribution of my `train` and `dev` sets (not so uncommon though).

NB: If you ignore run#1, run#2 to run#5 use the same datasets, the same batch sizes and, IMHO, decreasing learning rate between runs cannot explain the observed variations on itself.

**Comment by nicolaspanel** [comment]
@kdavis-mozilla @reuben  is it possible that dropout is not properly restored after the validation step ?
This could explain why validation loss is always lower on the first epoch.

**Comment by nicolaspanel** [comment]
> @kdavis-mozilla @reuben is it possible that dropout is not properly restored after the validation step ?

to confirm, I edited https://github.com/mozilla/DeepSpeech/blob/master/DeepSpeech.py#L279 to add
```py
    tf.summary.scalar(name='step_loss', tensor=avg_loss_across_towers, collections=['step_summaries'])
    for i, t in enumerate(dropout_rates):
        tf.summary.scalar(name='dropout_{}'.format(i), tensor=t, collections=['step_summaries'])
```
here is what I have in tensorboard:
![image](https://user-images.githubusercontent.com/2500584/54452980-a4411900-4756-11e9-897a-57a5d91c14a0.png)

_`dropout_0_1`, `dropout_1_1`, `dropout_2_1`, `dropout_3_1` , `dropout_4_1`, `dropout_5_1` correspond to `FLAGS.dropout_rate`, `FLAGS.dropout_rate2`, `FLAGS.dropout_rate3`, `FLAGS.dropout_rate4`, `FLAGS.dropout_rate5`, `FLAGS.dropout_rate6` (respectively) and I'm using `--dropout_rate 0.05 --dropout_rate6 0.5`_

=> [dropout_rates](https://github.com/mozilla/DeepSpeech/blob/master/DeepSpeech.py#L384) are properly set

**Comment by SamahZaro** [comment]
> to confirm, I edited https://github.com/mozilla/DeepSpeech/blob/master/DeepSpeech.py#L279 to add
> 
> ```python
>     tf.summary.scalar(name='step_loss', tensor=avg_loss_across_towers, collections=['step_summaries'])
>     for i, t in enumerate(dropout_rates):
>         tf.summary.scalar(name='dropout_{}'.format(i), tensor=t, collections=['step_summaries'])
> ```

dropout_rates variable is fixed to the values you pass as a parameter. This is not the right variable to track.

Did you successfully track the dropout probabilities per step? I need this part of code.

**Comment by kdavis-mozilla** [comment]
Closing due to lack of activity. Feel free to re-open if this is still a problem.
